ACCUMULO-4677 Sanitizing PathParam values in REST-based Monitor

glitch opened the pull request:
I took a swing at this one.  I wasn't fully sure what the bounds on some of the params should be (i.e. should minutes be bound between (0-59]  )?  What level of sanitization is sufficient, UrlEncoder kind  of stuff?  Take a look at the ParameterValidator class first if you're interested as it is the piece that's reused across the rest endpoints.

If you had something else in mind, feel free to decline or provide feedback and I can update.  Regards!
Note: Still running a local build and need to add further unit tests etc. if you're interested in this submission.

(Fails on RAT license check currently)

ctubbsii commented:
Oh, nice! Thanks for working on this @glitch ; I took a brief look and I think this is basically what we need. I didn't do a thorough review, though. I'll try to get to that soon, if nobody else does.

I think the main concern is that we don't allow input to be put back into the returned page in a way that poses a security risk. If a table can't be found, or a range of minutes doesn't work because it was input incorrectly, that's not so much a big deal, as long as it only breaks that particular HTTP request, and not Accumulo itself or the monitor state.

The main concern is probably being able to click a link to the monitor which causes the monitor to start executing javascript or something which was in the query or path parameters.

milleruntime commented on server/monitor/src/main/java/org/apache/accumulo/monitor/rest/trace/TracesResource.java:
```diff
@@ -176,12 +193,15 @@ public Void run() {
-    TraceList traces = new TraceList(id);
 
-    if (id == null) {
+    if (StringUtils.isEmpty(id)) {
```
I don't know what type of ID this is referring to but if it is a table or namespace, should create ID type objects.

milleruntime commented on server/monitor/src/main/java/org/apache/accumulo/monitor/rest/trace/TracesResource.java:
```diff
@@ -119,16 +125,27 @@ public Void run() {
+    final String type = ParameterValidator.sanitizeParameter(typeParameter);
+    
+    if (minutes >= 60 || minutes < 0) {
+      throw new Exception("minutes was not of range [0-59], was " + String.valueOf(minutes));
```
Should also use constants here.

milleruntime commented on server/monitor/src/main/java/org/apache/accumulo/monitor/rest/trace/TracesResource.java:
```diff
@@ -258,8 +278,8 @@ protected Range getRangeForTrace(long minutesSince) {
-    while (startHexTime.length() < endHexTime.length()) {
-      startHexTime = "0" + startHexTime;
+    if (startHexTime.length() < endHexTime.length()) {
+      StringUtils.leftPad(startHexTime, endHexTime.length(), '0');
```
Nice improvement!

milleruntime commented on server/monitor/src/main/java/org/apache/accumulo/monitor/rest/tservers/TabletServerResource.java:
```diff
@@ -141,28 +146,27 @@ public TabletServersRecovery getTserverRecovery() {
-  public TabletServerSummary getTserverDetails(@PathParam("address") String tserverAddr) throws Exception {
-
-    String tserverAddress = tserverAddr;
+  public TabletServerSummary getTserverDetails(@PathParam("address") String tserverAddress) throws Exception {
```
+1 I prefer full english words for parameter variable names (especially with Strings).  Unless you are using VIM as your IDE, abbreviation doesn't help anyone and only leads to confusion.  Granted, this example 99% of developers will know what "addr" means but who knows... there could be another variable to come along with the name tserverAdder or something.

glitch commented:
@milleruntime What's your feeling on using javax.validation.constraints right in the method signature?  Doing that I can pretty much eliminate the ParameterValidator class altogether.
i.e.
`@Path("show/{id}")
  @GET
  public TraceList getTracesType(@PathParam("id") @NotNull @Pattern(regexp = "\\w+") String id) throws Exception {
`
For int params:
`@DefaultValue("10") @PathParam("minutes") @NotNull @Min(0) @Max(2592000) int minutes`

It loses the single class encapsulation, but on the other hand you can really fine tune each method and it makes it self documenting.  It looks like support is built directly into Jersey for it.

Thoughts?

ctubbsii commented on server/monitor/src/main/java/org/apache/accumulo/monitor/util/ParameterValidator.java:
```diff
@@ -0,0 +1,48 @@
+   *           if the string cannot be encoded to UTF-8
+   */
+  public static String sanitizeParameter(String s) throws UnsupportedEncodingException {
+    return StringUtils.isEmpty(s) ? StringUtils.EMPTY : URLEncoder.encode(s, "UTF-8").trim();
```
I don't think this should be returning a URL-encoded version of the string. In most cases, we actually need to use the value after it has been decoded by Jersey. The only time we'd want to re-encode it is before display... but this method doesn't seem to be limited to only being used prior to display in the view.

It should also not throw an UnsupportedEncodingException. UTF-8 is always supported. This exception should be suppressed inside the method, instead of leaking out.

Also, `StandardCharsets.UTF_8.name()` should be used instead of `"UTF-8"`, whenever the String version of the charset is required.

glitch replied:
@ctubbsii I've been researching the annotation route using the javax.validation api which would remove the necessity of this class & methods altogether.

ctubbsii commented on server/monitor/src/main/java/org/apache/accumulo/monitor/util/ParameterValidator.java:
```diff
@@ -0,0 +1,48 @@
+    return StringUtils.isEmpty(s) ? StringUtils.EMPTY : URLEncoder.encode(s, "UTF-8").trim();
+  }
+
+  public static String sanitizeParameter(String s, String defaultValue) throws UnsupportedEncodingException {
```
Is this method needed? It seems redundant with the default value annotation.

glitch commented on server/monitor/src/main/java/org/apache/accumulo/monitor/rest/trace/TracesResource.java:
```diff
@@ -77,12 +81,12 @@
   @Path("summary/{minutes}")
   @GET
-  public RecentTracesList getTraces(@DefaultValue("10") @PathParam("minutes") int minutes) throws Exception {
+  public RecentTracesList getTraces(@DefaultValue("10") @PathParam("minutes") @Min(0) @Max(2592000) int minutes) throws Exception {
```
The annotations are part of the javax.validation api and Jersey 2.0+ has built in support for them.  According to the documentation the auto-discovery service will auto-enable them on the server without the need of manual configuration.  The reference implementation is the hibernate validator which will need to be explicitly added to the pom based on their documentation:

https://jersey.github.io/documentation/latest/bean-validation.html

I'm still reading up on it and it sounds like it could negate the need for the ParameterValidator class in this PR (which I'll end up removing).

ctubbsii replied:
Ah, interesting. So, it looks like there's a jersey extension which implements the validation for this API. Without the jersey-bean-validation, these are nothing more than ignored annotations. I'm curious what dependencies that might bring in. I don't think it'll be too bad, but need to be careful to include whatever we need in the assembly.

milleruntime replied:
Wouldn't defaulting to ignored annotation introduce a vulnerability in the validation? Say if an admin doesn't setup the monitor correctly or disables the auto-discovery.  I am wondering why we would ever want to disable the validation and can we prevent it from being disabled?

ctubbsii replied:
Presumably, we'd configure the monitor's servlet properly so that Jersey enforces them. This PR does not yet have those bits.

glitch commented:
Here is the link for the Jersey documentation surrounding parameter annotations.
https://jersey.github.io/documentation/latest/bean-validation.html

It's part of the JSR-303 spec and the reference implementation for the actual validation code is from hibernate-validator (org.hibernate) which looks to be pulled in as a transitive dependency.

milleruntime commented:
> @milleruntime What's your feeling on using javax.validation.constraints right in the method signature? 

I am on the fence with using external validation.  One one hand, it doesn't make sense to bring in a slew of dependencies just for a few annotations.  I have seen full web apps plagued by this in the past.  And since the monitor isn't the heart of Accumulo, I am more hesitant to include something like this.  On the other hand, if it is only one transitive dependency, I can definitely see the benefits it brings through custom code minimization and enhanced end point security, since this type of validation is quite common with web apps.

I am curious to see what type of errors are thrown by the validation.  I have seen frameworks like hibernate obscure error handling, especially at the front end.  Although, I don't think we have any error handling on the REST endpoints yet so perhaps that could be a follow on task.

glitch commented:
Ok a brief update.  For the transitive dependencies, this is what adding jersey-bean-validation brings along based on `mvn dependency:tree`
[INFO] +- org.glassfish.jersey.ext:jersey-bean-validation:jar:2.25.1:compile
[INFO] |  +- org.hibernate:hibernate-validator:jar:5.1.3.Final:compile
[INFO] |  |  +- org.jboss.logging:jboss-logging:jar:3.1.3.GA:compile
[INFO] |  |  \- com.fasterxml:classmate:jar:1.0.0:compile
[INFO] |  +- javax.el:javax.el-api:jar:2.2.4:compile
[INFO] |  \- org.glassfish.web:javax.el:jar:2.2.4:compile

I was able to get some basic unit testing set up on the REST endpoints testing for validation.  You can see that here: https://github.com/glitch/accumulo/commit/403b2adc2c6569eff6415ba6a64a2d95eb6f68d5

The annotation based constraints on parameters actually works pretty nicely.  As of now, failing a constraint will return a 400 response with a pretty generic message; however, there is the ability to tailor the responses/messages.  For a look at the annotations you can see what they look like here:
https://github.com/glitch/accumulo/commit/2e010deb8f40e9c6476ae4890d0b8de84c7f4274

I'm going to add some more unit testing and will merge those changes into the current pull-request.

milleruntime commented on pom.xml:
```diff
@@ -167,6 +167,11 @@
       </dependency>
       <dependency>
+        <groupId>com.fasterxml</groupId>
+        <artifactId>classmate</artifactId>
```
What is this used for?

glitch replied:
It's part of the transitive dependencies of using jersey's jersey-bean-validation and hibernate-validator default implementation (along with the javax.el deps, jboss-logging dep, etc.)

ctubbsii commented:
What's the status of this?

glitch commented:
Everything builds successfully for me locally.  I'm not sure what the deal is with the travis build, is there a way to kick it off again & have it run a full build?

Otherwise there did not appear to be any further requested changes and I believe it's good to go.

ctubbsii commented on server/monitor/src/test/java/org/apache/accumulo/monitor/view/WebViewsTest.java:
```diff
@@ -0,0 +1,151 @@
+  /**
+   * Silly stub to handle MessageBodyWriter for Hashmap. Registered in configure method and auto-wired by Jersey.
+   */
+  public static class HashMapWriter implements MessageBodyWriter<HashMap> {
```
Should specify generic type parameters here on HashMap (and also line 140 and 145) for type safety and to avoid introducing new warnings.

ctubbsii replied:
Reminder of these raw types warnings.

ctubbsii commented on assemble/src/main/assemblies/component.xml:
```diff
@@ -75,12 +77,17 @@
         <include>org.glassfish.jersey.media:jersey-media-jaxb</include>
         <include>org.glassfish.jersey.media:jersey-media-json-jackson</include>
+        <include>org.glassfish.web:javax.el</include>
+        <include>org.glassfish.web:javax.el-impl</include>
```
javax.el-impl isn't going to be included unless it's also a dependency of the assembly module, which does not appear to be the case.

ctubbsii commented:
@glitch I can't reproduce the compiler error in Travis, either. I re-kicked the build in Travis.

ctubbsii commented:
The only thing I can think of regarding the failure is that it has something to do with PowerMock. Do we need to use PowerMock for the WebViewsTest? Or will EasyMock on its own suffice?

glitch commented:
I needed PowerMock because EasyMock couldn't mock static methods (i.e. in the WebViewsTest needed to mock the Monitor.getContext() method to return a different mocked/stub context).

I can get rid of those specific tests using Powermock but I don't have anything to replace them with.

ctubbsii commented:
Hmm, I see. Well, maybe we can skip this test on Travis only?

glitch commented:
Ah, I got this to break locally with the right mvn incantation (was running the sunny day tests instead of the mvn clean verify -DskipITs that travis is running).  I'll see if I can get a workaround over Thanksgiving holiday; otherwise I can add an Ignore annotation or gut the PowerMock test.

ctubbsii commented:
The current branch is still internally consistent. The change in the master branch would only matter if one of those classes were somehow getting on the class path, which they shouldn't be. But, it is possible that the changes in master are more compatible with the mocking mechanism.

glitch commented:
Brief update:  The version of Jetty was changed recently to 9.3.21.v20170918.  This caused some internal dependency issues with jersey test framework (master's current jersey.version is 2.25.1 which carries a jetty dependency on version 9.2.14.v20151106).  There is no current jersey.version which matches master's jetty dependency (jersey skips from jetty 9.2.14 up to 9.4.3.v20170317 in 2.26-b03->4).

That aside I was able to rebase all commits, update code issues with class changes, and switch the jersey test provider from jetty to grizzly2 and all tests are passing now.

The last thing I have to do is test the full runtime locally.  Assuming it pans out, I'll force push over the Accumulo-4677 branch with the squashed commit for folks to look at again.  Thanks!

glitch commented:
Some good news, running locally the path validation looks to be working correctly (although the validation messages leave a bit to be desired).

That said I think I may have found some bugs in the monitor itself regarding some of the links (i.e. localhost:9995/tables/ displays tables but the hyperlink path ends with [object Object] instead of the table id).

In any case, assuming the Travis build is clean and no-one spots any other errors / assembly problems (fixed the el-impl one). I think this can be considered done for the initial validation capability.

glitch commented:
Yep, some of the hyperlinks for table paths are busted on master branch for the accumulo-monitor.

milleruntime commented:
> I think I may have found some bugs in the monitor itself regarding some of the links (i.e. localhost:9995/tables/ displays tables but the hyperlink path ends with [object Object]

+1 For the new ticket you created.  I noticed this while working on ACCUMULO-4739 but forgot to create a ticket.  Thanks!

milleruntime commented:
I thought I saw a comment from @ctubbsii but I don't see it here? Perhaps it was lost in your last rebase... either way, if he is good with this PR, I can merge it.

ctubbsii commented on pom.xml:
```diff
@@ -703,6 +738,12 @@
+        <groupId>org.glassfish.jersey.test-framework.providers</groupId>
+        <artifactId>jersey-test-framework-provider-grizzly2</artifactId>
+        <version>${jersey.version}</version>
+        <scope>test</scope>
```
We typically don't set the scope in the parent POM. We manage the dependency versions in the parent POM, and choose the scope in the module where it is used.

ctubbsii commented on server/monitor/pom.xml:
```diff
@@ -123,6 +127,32 @@
+    <dependency>
+      <groupId>org.glassfish.jersey.test-framework.providers</groupId>
+      <artifactId>jersey-test-framework-provider-grizzly2</artifactId>
+      <version>${jersey.version}</version>
```
Should not be overriding version from parent POM dependencyManagement section here.

glitch commented:
...and more conflicts & test failures because the underlying WebViews class changed again, lol.

ctubbsii commented on server/monitor/src/test/java/org/apache/accumulo/monitor/view/WebViewsTest.java:
```diff
@@ -130,6 +130,7 @@ public void testGetTracesSummaryValidationConstraint() {
   /**
    * Silly stub to handle MessageBodyWriter for Hashmap. Registered in configure method and auto-wired by Jersey.
    */
+  @SuppressWarnings("rawtypes")
```
`<?,?>` as the generic type would be much preferred over suppressing the warning. It's always better to be explicit, and I'm not a fan of instructing the compiler to ignore warnings when there's an easy fix.

Also, I just noticed that this test creates an external web service to test against. As such, this test should be renamed to be `WebViewsIT`, because it's an integration test, and not just a unit test.

And, don't worry about the conflicts with master. I can take care of that for you (going to rebase before merge). I can also address these last few minor issues, too, during the rebase/merge process.

glitch replied:
RE: IntegrationTests which @Category class should this annotation point at?

ctubbsii replied:
Ugh, I forgot about categories. Don't worry about that. I'll deal with it as a follow-on issue. It's fine as it is for now.

asfgit commented:
Can one of the admins verify this patch?

glitch commented:
Ok, I renamed the test class to WebViewsIT and moved it into the accumulo-test module under its own package (org.apache.accumulo.test.monitor) and annotated with the SunnyDayTests @Category annotation.  I also updated the various pom dependencies accordingly (basically moving the test scope deps from accumulo-monitor pom to accumulo-test pom etc.)  The issue with the rawtypes versus generic types has also been fixed on the internal test HashMapWriter stub.

I'm still waiting for the ITs to finish locally but I think he has everything folks wanted.

Let me know if there is something else, I don't mind fixing it and getting it in the correct spot.  I also don't mind if someone wants to touch it up later.

milleruntime commented:
@glitch Sorry for the conflicts, I had wrapped up ACCUMULO-4739 yesterday but missed a few updates. My changes to WebViews should be isolated to the new method I created though.  I will put off the rest of the changes I wanted to make to the Monitor until we can get this merged up.

milleruntime commented:
@glitch If you want me to fix the conflicts for you, just let me know.  I can check out the PR, squash and merge it all up since your work looks complete.

milleruntime commented:
I think i fixed the dependency issues with test pom with d3cf72f but the WevViewsiT is still failing.

milleruntime commented:
The Travis build from my commit last night passed: https://travis-ci.org/apache/accumulo/builds/308707934
But it didn't run the ITs.  

When I run WebViewsIT by itself (mvn clean verify -Dit.test=WebViewsIT -Dtest=foo -Dfindbugs.skip -Dcheckstyle.skip -DfailIfNoTests=false) I get the wrong HTTP status returned in 2 of the tests:
```
WebViewsIT.testGetTablesConstraintViolations:85 should return status 400 expected:<400> but was:<500>
WebViewsIT.testGetTracesSummaryValidationConstraint:129 should return status 400 expected:<400> but was:<500>
```

glitch commented:
Ok, this looks like it's passing locally now.  The culprit was some missing dependencies for the bean validation stuff to wire itself correctly.

This worked:  `mvn clean verify -Dit.test=WebViewsIT -Dtest=foo -Dfindbugs.skip -Dcheckstyle.skip -DfailIfNoTests=false`

As well as `mvn verify -Psunny`

milleruntime commented:
Thanks @glitch !  I will merge this

milleruntime commented:
@glitch I noticed that the getTables validation will fail on all of the accumulo table IDs.  The ALPHA_NUM_REGEX doesn't like "+" or the "!0" ID.  

I am not sure what the best way to fix this would be.  We could either add a more complicated regex that will allow the special IDs or change the REST interface to use table names instead of IDs.  Thoughts?

milleruntime commented:
Created https://issues.apache.org/jira/browse/ACCUMULO-4756 for this issue
